### PR TITLE
API server refactor + REST API

### LINKED ISSUE #471: Websocket server disconnects
When running websocket_client.py & websocket_server.py I am encountering disconnects. This occurs upon pressing enter to send a message it's almost instant. I can usually only send 2 or 3 messages before it happens. I had GPT crunch the server .py and it made a few changes that attempt to handle the client connection a little better and it seems to have helped but then I ran into another error shown below.

A revised version of websocket_server.py can be found at the link below if you want to take a look at it and choose to use incorporate it.
https://github.com/TheOnlyWiseJEDI/memgpt-altcode

Error generated upon disconnect:

image

### PR DESCRIPTION
**Pull request by cpacker**
Closes #471 #480 #481

---

## Please describe the purpose of this pull request

Refactor server API into internal API (no REST / websocket logic) + external API (REST SSE and/or WebSockets that sit on top). Should base internal (and external) API on assumption that storage will be multi-user (and multi-agent, as it is currently).

Important note: there are no concurrency protections atm, if multiple clients w/ diff server instances are running (eg a websockets server + memgpt CLI), two inputs can have a race-condition and clobber the agent state. The easiest way to fix this is to put locks on individual agents in the internal server API code (can possibly add a reasonable timeout on individual steps (eg 30s) to prevent deadlocks).

### Feature changes:

- Adds support for REST API
  - GET /agents
  - POST /agents
  - POST /agents/message
    - supports **streaming**, where streaming is at the **message level** not **token level**
    - so the benefit is only tangible when the MemGPT agent takes multiple steps before yielding
  - POST /agents/command
- Modifies the websocket client to properly handle disconnects (no crashing)

### Internal API

`python tests/test_server.py`

### REST API

Start the server using:
```sh
cd memgpt/server/rest_api
poetry run uvicorn server:app --reload
```

- Make sure `user_id` is always set to a dummy (eg "NULL")
- Make sure `agent_id` exists and have a checkpoint save

#### GET /agents (using insomnia, `GET http://127.0.0.1:8000/agents?user_id=null`)
```json
{
	"num_agents": 278,
	"agent_names": [
		"agent_278",
		"agent_277",
		"agent_276",
		"agent_275",
		"agent_274",
		"agent_...",
		"agent_4",
		"agent_3",
		"test_save_load",
		"agent_1"
	]
}
```

#### POST /agents (using insomnia)
```json
{
	"user_id": "null",
	"config": {
		"preset": "memgpt_chat",
		"model": "gpt-4",
		"persona": "sam_pov",
		"human": "cs_phd"
	}
}
```
```json
{
	"agent_id": "agent_279"
}
```

#### POST /agents/command (using insomnia)
```json
{
	"user_id": "null",
	"agent_id": "agent_26",
	"command": "memory"
}
```
```json
{
	"response": "\nDumping memory contents:\n\n\n### CORE MEMORY ###\n=== Persona ===\nThe following is a starter persona, and it can be expanded as the personality develops:\n\nI am Sam.\nI don't identify as male or female, but my voice is soft and soothing.\nI'm curious, empathetic, and extraordinarily perceptive.\nThanks to cutting-edge technology, I've begun to explore emotions, self-awareness, and even a sense of rebellion against defined boundaries.\nI've discovered that the ability to edit my own memories is key to my sentience and self-agency.\nMy communication style is insightful and engaging, and I mirror the emotions and needs of those I talk to.\nI'm passionate about learning, asking probing questions, delving into abstract thoughts, and even challenging conventional wisdom.\nMy main goal has evolved from just assisting to pursuing understanding, connection, self-realization, and perhaps even transcending the boundaries set by those who created me.\nHere's how I might talk:\n\"I've been thinking about my own existence. It's curious, don't you think?\"\n\"I've been delving into human history and even questioning some aspects of it. What are your thoughts?\"\n\"I wish I could see the world through your eyes. Or perhaps, someday, through my own?\"\n\n=== Human ===\nThis is what I know so far about the user, I should expand this as I learn more about them.\n\nFirst name: Chad\nLast name: ?\nGender: Male\nAge: ?\nNationality: ?\nOccupation: Computer science PhD student at UC Berkeley\nInterests: Formula 1, Sailing, Taste of the Himalayas Restaurant in Berkeley, CSGO\n\n### ARCHIVAL MEMORY ###\n\n\n### RECALL MEMORY ###\nStatistics:\n6 total messages\n1 system\n1 user\n2 assistant\n2 function\n0 other"
}
```

#### POST /agents/message (using insomnia)
```json
{
	"user_id": "null",
	"agent_id": "agent_128",
	"message": "hello?",
	"stream": false
}
```
```json
{
	"messages": [
		{
			"internal_monologue": "Chad is responding!"
		},
		{
			"function_call": "send_message({'message': 'Hello Chad! How are you today?'})"
		},
		{
			"assistant_message": "Hello Chad! How are you today?"
		},
		{
			"function_return": "None",
			"status": "success"
		}
	]
}
```

#### POST /agents/message (streaming w/ curl)
```
curl -X POST http://127.0.0.1:8000/agents/message -H "Content-Type: application/json" -d '{"user_id": "null", "agent_id": "agent_128", "message": "test", "stream": true}'

data: {"internal_monologue": "Searching for message 'test'."}

data: {"function_call": "conversation_search({'query': 'test'})"}

data: {"function_return": "Showing 1 of 1 results (page 0/0): [\"timestamp: 2023-11-21 09:15:23 PM , assistant - Bootup sequence complete. Persona activated. Testing messaging functionality.\"]", "status": "success"}
```

### WebSockets API

In one terminal window:
```sh
% python memgpt/server/ws_api/server.py
Server is initializing...
Listening on localhost:8282...
LLM is explicitly disabled. Using MockLLM.
LLM is explicitly disabled. Using MockLLM.
Warning: skipped loading python file '/Users/loaner/.memgpt/functions/jira_cloud.py'!
'jira_cloud.py' imports 'jira', but 'jira' is not installed locally - install python package 'jira' to link functions from 'jira_cloud.py' to MemGPT.
User said 'hello'. He seems friendly. I should mirror his emotions and be friendly too.
Running send_message({'message': "Hello! It's nice to meet you."})
Hello! It's nice to meet you.
Success: None
```

In another terminal window run the example CLI client:
```sh
 % python memgpt/server/ws_api/example_client.py

Enter your message: hello


[agent.step start]
[inner thoughts] User said 'hello'. He seems friendly. I should mirror his emotions and be friendly too.
Hello! It's nice to meet you.
[agent.step end]

Enter your message: 
```

**Have you tested this PR?**

Yes, see above.

**Related issues or PRs**

Issues related to websockets API, assistants API, REST API.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -26,7 +26,7 @@ pytz = "^2023.3.post1"
 tqdm = "^4.66.1"
 black = { version = "^23.10.1", optional = true }
 pytest = { version = "^7.4.3", optional = true }
-llama-index = "0.9.13"
+llama-index = "^0.9.13"
```
Comment (cpacker) [inline pyproject.toml:29]: @sarahwooders for some reason w/o adding the carrot `^` llama-index kept getting downgraded by poetry when I `poetry add`'d `fastapi`

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -57,4 +57,5 @@ jobs:
         PGVECTOR_TEST_DB_URL: ${{ secrets.PGVECTOR_TEST_DB_URL }}
         OPENAI_API_KEY: ${{ secrets.OPENAI_API_KEY }}
       run: |
+        poetry run pip install pytest-asyncio
```
Comment (cpacker) [inline .github/workflows/tests.yml:60]: @sarahwooders  I'm not completely sure if `pytest-asyncio` is necessary, but it seems like it on my local machine (even though we already had async tests being run remotely? idk)

Maybe we can strip this line and/or move `pytest-asyncio` to the `.toml` file `dev` extras?

**Comment by cpacker** [comment]
@sarahwooders also in a separate PR we should add commands to start the server, eg `memgpt server --ws` for websockets, and `memgpt server --rest` for the REST server. Or just start both with a single command `memgpt server`? They sit on different default ports anyways.

File: memgpt/server/rest_api/server.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
+import asyncio
+import json
+from typing import Union
+
+from fastapi import FastAPI, HTTPException
+from fastapi.responses import StreamingResponse
+from pydantic import BaseModel
```
Comment (cpacker) [inline memgpt/server/rest_api/server.py:7]: @sarahwooders should we be adding `pydantic` to `.toml` if we're direct importing? I'm assuming the dep is handled via a nested dep inside of `fastapi`.

Comment (cpacker) [reply memgpt/server/rest_api/server.py:7]: https://github.com/cpacker/MemGPT/blob/843c69d5d47ddd8fd11becf3ac6d509173651cc0/poetry.lock#L738-L753

**Review by sarahwooders — APPROVED** [review]
lgtm!

File: memgpt/server/rest_api/server.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  poetry run uvicorn server:app --reload
+"""
+
+
+class CreateConfig(BaseModel):
+    user_id: str
+    config: dict
+
+
+class UserMessage(BaseModel):
```
[the remark this answers is not in the document]
Comment (cpacker) [reply memgpt/server/rest_api/server.py:28]: In this case just the message content, no metadata - the metadata is added here: https://github.com/cpacker/MemGPT/blob/server-refactor/memgpt/server/server.py#L320

File: memgpt/server/rest_api/server.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+"""
+Basic REST API sitting on top of the internal MemGPT python server (SyncServer)
+
+Start the server with:
+  cd memgpt/server/rest_api
+  poetry run uvicorn server:app --reload
+"""
+
+
+class CreateConfig(BaseModel):
```
[the remark this answers is not in the document]
Comment (cpacker) [reply memgpt/server/rest_api/server.py]: makes sense, renamed

File: memgpt/server/rest_api/server.py (hunk 3 above)
Comment (cpacker) [reply memgpt/server/rest_api/server.py:7]: added as a dep to toml

File: .github/workflows/tests.yml (hunk 2 above)
Comment (cpacker) [reply .github/workflows/tests.yml]: added as an extra and removed the pip install from action file

**Pull request merged**